application/app/controllers/dataverse/dataverses_controller.rb

```diff
@@ -0,0 +1,36 @@
+class Dataverse::DataversesController < ApplicationController
+  include LoggingCommon
+
+  before_action :get_dv_full_hostname
+  before_action :init_service
+
+  def index
+    @dataverse = @service.find_dataverse_by_id(":root")
+    @search_result = @service.search_dataverse_items(":root")
+  end
+
+  def show
+    @dataverse = @service.find_dataverse_by_id(params[:id])
+    @search_result = @service.search_dataverse_items(params[:id])
+    render :index
+  end
+
+  private
+
+  def get_dv_full_hostname
+    if params[:dv_hostname]
+      hostname = params[:dv_hostname]
+      scheme = params[:dv_scheme] || "https"
+      port = params[:dv_port] || 443
+      @dataverse_url = URI.parse(scheme + "://" + hostname + ":" + port.to_s).to_s
```

Comment:
I think this is used elsewhere, can be create a utility class to create the DV URL

Reply:
yes, it is duplicated. let me refactor it

Reply:
done